networkit/cpp/distance/APSP.cpp

```diff
@@ -15,21 +16,27 @@ namespace NetworKit {
 APSP::APSP(const Graph &G) : Algorithm(), G(G) {}
 
 void APSP::run() {
-    std::vector<edgeweight> distanceVector(G.upperNodeIdBound(), 0.0);
-    distances.resize(G.upperNodeIdBound(), distanceVector);
-    if (G.isWeighted()) {
-        G.parallelForNodes([&](node u) {
-            Dijkstra dijk(G, u);
-            dijk.run();
-            distances[u] = dijk.getDistances();
-        });
-    } else {
-        G.parallelForNodes([&](node u) {
-            BFS bfs(G, u);
-            bfs.run();
-            distances[u] = bfs.getDistances();
-        });
+    count n = G.upperNodeIdBound();
+    std::vector<edgeweight> distanceVector(n, 0.0);
+    distances.resize(n, distanceVector);
+
+    count nThreads = omp_get_max_threads();
+    sssps.reserve(nThreads);
+    for (count i = 0; i < nThreads; ++i) {
+        if (G.isWeighted())
+            sssps.emplace_back(new Dijkstra(G, 0, false));
+        else
+            sssps.emplace_back(new BFS(G, 0, false));
     }
```

Comment:
As a minor comment, it might make sense to a) resize sssps to the desired size b) open the parallel region and then c) allocate the sssps in paralle, i.e., allocate with each thread the entry for `omp_get_thread_num()`. This would ensure that at least the SSSP-object is allocated in the correct numa region if threads are pinned. However, as this is not considered in most of the code in NetworKit anyway, this might not make much sense here.

Reply:
That's true in principle. OTOH I don't know how well the glibc memory allocator handles NUMA. If manipulation of the underlying data structures is protected by a global lock, it might not be so much faster.

But it's a low-effort optimization, so yes, why not do it.

Reply:
I implemented this optimization, I'll push it as soon as we resolve https://github.com/kit-parco/networkit/pull/327#discussion_r291573908